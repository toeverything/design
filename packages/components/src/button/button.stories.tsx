import { InformationIcon } from '@blocksuite/icons';
import type { Meta, StoryFn } from '@storybook/react';

import { Button, type ButtonProps } from './button';
export default {
  title: 'UI/Button',
  component: Button,
  argTypes: {
    onClick: () => console.log('Click button'),
  },
} as Meta<ButtonProps>;

const Template: StoryFn<ButtonProps> = args => <Button {...args} />;

export const Default = Template.bind(undefined);
Default.args = {
  type: 'default',
  children: 'This is a default button',
  icon: <InformationIcon />,
};

export const Primary = Template.bind(undefined);
Primary.args = {
  type: 'primary',
  children: 'Content',
  icon: <InformationIcon />,
};

export const Disabled = Template.bind(undefined);
Disabled.args = {
  disabled: true,
  children: 'This is a disabled button',
};

export const LargeSizeButton = Template.bind(undefined);
LargeSizeButton.args = {
  size: 'large',
  children: 'This is a large button',
};

export const ExtraLargeSizeButton = Template.bind(undefined);
ExtraLargeSizeButton.args = {
  size: 'extraLarge',
  children: 'This is a extra large button',
};
