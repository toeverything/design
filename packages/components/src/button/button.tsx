import clsx from 'clsx';
import {
  type FC,
  forwardRef,
  type HTMLAttributes,
  type PropsWithChildren,
  type ReactElement,
  useMemo,
} from 'react';

import { Loading } from '../loading';
import { button, buttonIcon } from './style.css';
export type ButtonType =
  | 'default'
  | 'primary'
  | 'plain'
  | 'error'
  | 'warning'
  | 'success'
  | 'processing';
export type ButtonSize = 'default' | 'large' | 'extraLarge';
export type ButtonProps = PropsWithChildren &
  Omit<HTMLAttributes<HTMLButtonElement | HTMLDivElement>, 'type'> & {
    type?: ButtonType;
    disabled?: boolean;
    icon?: ReactElement;
    iconPosition?: 'start' | 'end';
    shape?: 'default' | 'round' | 'circle';
    block?: boolean;
    size?: ButtonSize;
    loading?: boolean;
    withoutHoverStyle?: boolean;
  };
const defaultProps = {
  type: 'default',
  disabled: false,
  shape: 'default',
  size: 'default',
  iconPosition: 'start',
  loading: false,
  withoutHoverStyle: false,
};

const ButtonIcon: FC<ButtonProps> = props => {
  const {
    size,
    icon,
    iconPosition = 'start',
    children,
    type,
    loading,
    ...otherProps
  } = {
    ...defaultProps,
    ...props,
  };
  const onlyIcon = icon && !children;
  return (
    <div
      {...otherProps}
      className={clsx(buttonIcon, {
        'color-white': type !== 'default' && type !== 'plain',
        large: size === 'large',
        extraLarge: size === 'extraLarge',
        end: iconPosition === 'end' && !onlyIcon,
        start: iconPosition === 'start' && !onlyIcon,
        loading,
      })}
    >
      {icon}
    </div>
  );
};
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      children,
      type,
      disabled,
      shape,
      size,
      icon: propsIcon,
      iconPosition,
      block,
      loading,
      withoutHoverStyle,
      className,
      ...otherProps
    } = {
      ...defaultProps,
      ...props,
    };

    const icon = useMemo(() => {
      if (loading) {
        return <Loading />;
      }
      return propsIcon;
    }, [propsIcon, loading]);

    return (
      <button
        {...otherProps}
        ref={ref}
        className={clsx(
          button,
          {
            primary: type === 'primary',
            plain: type === 'plain',
            error: type === 'error',
            warning: type === 'warning',
            success: type === 'success',
            processing: type === 'processing',
            large: size === 'large',
            extraLarge: size === 'extraLarge',
            disabled,
            circle: shape === 'circle',
            round: shape === 'round',
            block,
            loading,
            'without-hover': withoutHoverStyle,
          },
          className
        )}
        disabled={disabled}
        data-disabled={disabled}
      >
        {icon && iconPosition === 'start' ? (
          <ButtonIcon {...props} icon={icon} />
        ) : null}
        <span>{children}</span>
        {icon && iconPosition === 'end' ? (
          <ButtonIcon {...props} icon={icon} />
        ) : null}
      </button>
    );
  }
);
Button.displayName = 'Button';
export default Button;
