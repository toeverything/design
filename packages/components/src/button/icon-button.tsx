import clsx from 'clsx';
import type { HTMLAttributes, PropsWithChildren } from 'react';
import { forwardRef, type ReactElement } from 'react';

import { Loading } from '../loading';
import type { ButtonType } from './button';
import { iconButton } from './style.css';

export type IconButtonSize = 'default' | 'large' | 'small' | 'extraSmall';
export type IconButtonProps = PropsWithChildren &
  Omit<HTMLAttributes<HTMLButtonElement>, 'type'> & {
    type?: ButtonType;
    disabled?: boolean;
    size?: IconButtonSize;
    loading?: boolean;
    withoutPadding?: boolean;
    active?: boolean;
    withoutHoverStyle?: boolean;
    icon?: ReactElement;
  };
const defaultProps = {
  type: 'plain',
  disabled: false,
  size: 'default',
  loading: false,
  withoutPadding: false,
  active: false,
  withoutHoverStyle: false,
};

export const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  (props, ref) => {
    const {
      type,
      size,
      withoutPadding,
      children,
      disabled,
      loading,
      active,
      withoutHoverStyle,
      icon: propsIcon,
      className,
      ...otherProps
    } = {
      ...defaultProps,
      ...props,
    };

    return (
      <button
        ref={ref}
        className={clsx(
          iconButton,
          {
            'without-padding': withoutPadding,

            primary: type === 'primary',
            plain: type === 'plain',
            error: type === 'error',
            warning: type === 'warning',
            success: type === 'success',
            processing: type === 'processing',

            large: size === 'large',
            small: size === 'small',
            'extra-small': size === 'extraSmall',

            disabled,
            loading,
            active,
            'without-hover': withoutHoverStyle,
          },
          className
        )}
        disabled={disabled}
        data-disabled={disabled}
        {...otherProps}
      >
        {loading ? <Loading /> : children || propsIcon}
      </button>
    );
  }
);

IconButton.displayName = 'IconButton';
export default IconButton;
