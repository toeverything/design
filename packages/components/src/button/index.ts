export * from './button.tsx';
export * from './icon-button.tsx';
