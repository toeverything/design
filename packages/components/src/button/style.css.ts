import { globalStyle, style } from '@vanilla-extract/css';

export const button = style({
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  userSelect: 'none',
  touchAction: 'manipulation',
  outline: '0',
  border: '1px solid',
  padding: '0 18px',
  borderRadius: '8px',
  fontSize: 'var(--affine-font-xs)',
  fontWeight: 500,
  transition: 'all .3s',
  ['WebkitAppRegion' as string]: 'no-drag',
  cursor: 'pointer',

  // changeable
  height: '28px',
  background: 'var(--affine-white)',
  borderColor: 'var(--affine-border-color)',
  color: 'var(--affine-text-primary-color)',

  selectors: {
    '&.text-bold': {
      fontWeight: 600,
    },
    '&:not(.without-hover):hover': {
      background: 'var(--affine-hover-color)',
    },
    '&.disabled': {
      opacity: '.4',
      cursor: 'default',
      color: 'var(--affine-disable-color)',
      pointerEvents: 'none',
    },
    '&.loading': {
      cursor: 'default',
      color: 'var(--affine-disable-color)',
      pointerEvents: 'none',
    },
    '&.disabled:not(.without-hover):hover, &.loading:not(.without-hover):hover':
      {
        background: 'inherit',
      },

    '&.block': { display: 'flex', width: '100%' },

    '&.circle': {
      borderRadius: '50%',
    },
    '&.round': {
      borderRadius: '14px',
    },
    // size
    '&.large': {
      height: '32px',
      fontSize: 'var(--affine-font-base)',
      fontWeight: 600,
    },
    '&.round.large': {
      borderRadius: '16px',
    },
    '&.extraLarge': {
      height: '40px',
      fontSize: 'var(--affine-font-base)',
      fontWeight: 700,
    },
    '&.round.extraLarge': {
      borderRadius: '20px',
    },

    // type
    '&.plain': {
      color: 'var(--affine-text-primary-color)',
      borderColor: 'transparent',
      background: 'transparent',
    },

    '&.primary': {
      color: 'var(--affine-pure-white)',
      background: 'var(--affine-primary-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: 'var(--affine-button-inner-shadow)',
    },
    '&.primary:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-primary-color)',
    },
    '&.primary.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.primary.disabled:not(.without-hover):hover': {
      background: 'var(--affine-primary-color)',
    },

    '&.error': {
      color: 'var(--affine-pure-white)',
      background: 'var(--affine-error-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: 'var(--affine-button-inner-shadow)',
    },
    '&.error:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-error-color)',
    },
    '&.error.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.error.disabled:not(.without-hover):hover': {
      background: 'var(--affine-error-color)',
    },

    '&.warning': {
      color: 'var(--affine-pure-white)',
      background: 'var(--affine-warning-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: 'var(--affine-button-inner-shadow)',
    },
    '&.warning:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-warning-color)',
    },
    '&.warning.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.warning.disabled:not(.without-hover):hover': {
      background: 'var(--affine-warning-color)',
    },

    '&.success': {
      color: 'var(--affine-pure-white)',
      background: 'var(--affine-success-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: 'var(--affine-button-inner-shadow)',
    },
    '&.success:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-success-color)',
    },
    '&.success.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.success.disabled:not(.without-hover):hover': {
      background: 'var(--affine-success-color)',
    },

    '&.processing': {
      color: 'var(--affine-pure-white)',
      background: 'var(--affine-processing-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: 'var(--affine-button-inner-shadow)',
    },
    '&.processing:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-processing-color)',
    },
    '&.processing.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.processing.disabled:not(.without-hover):hover': {
      background: 'var(--affine-processing-color)',
    },
  },
});

globalStyle(`${button} > span`, {
  // flex: 1,
  lineHeight: 1,
  padding: '0 4px',
});

export const buttonIcon = style({
  flexShrink: 0,
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  color: 'var(--affine-icon-color)',
  fontSize: '16px',
  width: '16px',
  height: '16px',
  selectors: {
    '&.start': {
      marginRight: '4px',
    },
    '&.end': {
      marginLeft: '4px',
    },
    '&.large': {
      fontSize: '20px',
      width: '20px',
      height: '20px',
    },
    '&.extraLarge': {
      fontSize: '20px',
      width: '20px',
      height: '20px',
    },
    '&.color-white': {
      color: 'var(--affine-pure-white)',
    },
  },
});

export const iconButton = style({
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  userSelect: 'none',
  touchAction: 'manipulation',
  outline: '0',
  border: '1px solid',
  borderRadius: '4px',
  transition: 'all .3s',
  ['WebkitAppRegion' as string]: 'no-drag',
  cursor: 'pointer',

  // changeable
  width: '24px',
  height: '24px',
  fontSize: '20px',
  color: 'var(--affine-text-primary-color)',
  borderColor: 'var(--affine-border-color)',
  selectors: {
    '&.without-padding': {
      margin: '-2px',
    },
    '&.active': {
      color: 'var(--affine-primary-color)',
    },

    '&:not(.without-hover):hover': {
      background: 'var(--affine-hover-color)',
    },
    '&.disabled': {
      opacity: '.4',
      cursor: 'default',
      color: 'var(--affine-disable-color)',
      pointerEvents: 'none',
    },
    '&.loading': {
      cursor: 'default',
      color: 'var(--affine-disable-color)',
      pointerEvents: 'none',
    },
    '&.disabled:not(.without-hover):hover, &.loading:not(.without-hover):hover':
      {
        background: 'inherit',
      },

    // size
    '&.large': {
      width: '32px',
      height: '32px',
      fontSize: '24px',
    },
    '&.large.without-padding': {
      margin: '-4px',
    },
    '&.small': { width: '20px', height: '20px', fontSize: '16px' },
    '&.extra-small': { width: '16px', height: '16px', fontSize: '12px' },

    // type
    '&.plain': {
      color: 'var(--affine-icon-color)',
      borderColor: 'transparent',
      background: 'transparent',
    },
    '&.plain.active': {
      color: 'var(--affine-primary-color)',
    },

    '&.primary': {
      color: 'var(--affine-white)',
      background: 'var(--affine-primary-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: '0px 1px 2px 0px rgba(255, 255, 255, 0.25) inset',
    },
    '&.primary:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-primary-color)',
    },
    '&.primary.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.primary.disabled:not(.without-hover):hover': {
      background: 'var(--affine-primary-color)',
    },

    '&.error': {
      color: 'var(--affine-white)',
      background: 'var(--affine-error-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: '0px 1px 2px 0px rgba(255, 255, 255, 0.25) inset',
    },
    '&.error:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-error-color)',
    },
    '&.error.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.error.disabled:not(.without-hover):hover': {
      background: 'var(--affine-error-color)',
    },

    '&.warning': {
      color: 'var(--affine-white)',
      background: 'var(--affine-warning-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: '0px 1px 2px 0px rgba(255, 255, 255, 0.25) inset',
    },
    '&.warning:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-warning-color)',
    },
    '&.warning.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.warning.disabled:not(.without-hover):hover': {
      background: 'var(--affine-warning-color)',
    },

    '&.success': {
      color: 'var(--affine-white)',
      background: 'var(--affine-success-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: '0px 1px 2px 0px rgba(255, 255, 255, 0.25) inset',
    },
    '&.success:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-success-color)',
    },
    '&.success.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.success.disabled:not(.without-hover):hover': {
      background: 'var(--affine-success-color)',
    },

    '&.processing': {
      color: 'var(--affine-white)',
      background: 'var(--affine-processing-color)',
      borderColor: 'var(--affine-black-10)',
      boxShadow: '0px 1px 2px 0px rgba(255, 255, 255, 0.25) inset',
    },
    '&.processing:not(.without-hover):hover': {
      background:
        'linear-gradient(0deg, rgba(0, 0, 0, 0.04) 0%, rgba(0, 0, 0, 0.04) 100%), var(--affine-processing-color)',
    },
    '&.processing.disabled': {
      opacity: '.4',
      cursor: 'default',
    },
    '&.processing.disabled:not(.without-hover):hover': {
      background: 'var(--affine-processing-color)',
    },
  },
});
