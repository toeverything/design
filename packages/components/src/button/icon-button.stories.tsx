import { InformationIcon } from '@blocksuite/icons';
import type { Meta, StoryFn } from '@storybook/react';

import { IconButton, type IconButtonProps } from './icon-button';
export default {
  title: 'UI/IconButton',
  component: IconButton,
  argTypes: {
    onClick: () => console.log('Click button'),
  },
} as Meta<IconButtonProps>;

const Template: StoryFn<IconButtonProps> = args => <IconButton {...args} />;

export const Plain = Template.bind(undefined);
Plain.args = {
  children: <InformationIcon />,
};

export const Priamry = Template.bind(undefined);
Priamry.args = {
  type: 'primary',
  icon: <InformationIcon />,
};

export const Disabled = Template.bind(undefined);
Disabled.args = {
  disabled: true,
  icon: <InformationIcon />,
};
export const ExtraSmallSizeButton = Template.bind(undefined);
ExtraSmallSizeButton.args = {
  size: 'extraSmall',
  icon: <InformationIcon />,
};
export const SmallSizeButton = Template.bind(undefined);
SmallSizeButton.args = {
  size: 'small',
  icon: <InformationIcon />,
};
export const LargeSizeButton = Template.bind(undefined);
LargeSizeButton.args = {
  size: 'large',
  icon: <InformationIcon />,
};
