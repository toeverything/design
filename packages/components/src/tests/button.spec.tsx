import renderer from 'react-test-renderer';
import { expect, test } from 'vitest';

import { Button } from '../button';
import toJson from './utils/to-json';

test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <Button type="primary">This is primary button</Button>
  );
  let tree = toJson(component);
  expect(tree).toMatchSnapshot();

  tree.props.type = 'primary';

  tree = toJson(component);
  expect(tree).toMatchSnapshot();

  tree.props.type = 'plain';

  tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
