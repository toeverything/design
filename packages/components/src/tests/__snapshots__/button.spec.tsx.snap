// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link changes the class when hovered 1`] = `
<button
  className="style_button__q047v90 primary"
  data-disabled={false}
  disabled={false}
>
  <span>
    This is primary button
  </span>
</button>
`;

exports[`Link changes the class when hovered 2`] = `
<button
  className="style_button__q047v90 primary"
  data-disabled={false}
  disabled={false}
>
  <span>
    This is primary button
  </span>
</button>
`;

exports[`Link changes the class when hovered 3`] = `
<button
  className="style_button__q047v90 primary"
  data-disabled={false}
  disabled={false}
>
  <span>
    This is primary button
  </span>
</button>
`;
